### PR TITLE
 Remove lock from AggregateDependenciesSnapshotProvider 

### PR DESCRIPTION
**Pull request by drewnoakes**
The collection protected by this lock changes infrequently, yet it is read from very regularly during dependency updates. It's a good fit for opportunistic interlocked operations on an immutable dictionary.

Removing this lock also removes the potential for contention across projects, as `AggregateDependenciesSnapshotProvider` brokers data from all projects in the solution.

### CODE REVIEW

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs [hunk 1]
```diff
@@ -42,42 +45,49 @@ public void RegisterSnapshotProvider(IDependenciesSnapshotProvider snapshotProvi
...
             }
 
-            lock (_snapshotProviders)
-            {
-                _snapshotProviders[snapshotProvider.CurrentSnapshot.ProjectPath] = snapshotProvider;
-                snapshotProvider.SnapshotRenamed += OnSnapshotRenamed;
-                snapshotProvider.SnapshotChanged += OnSnapshotChanged;
-                snapshotProvider.SnapshotProviderUnloading += OnSnapshotProviderUnloading;
-            }
+            ImmutableInterlocked.Update(
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs]: ❔  Why not `ImmutableInterlocked.TryAdd`?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs [hunk 2]
```diff
@@ -42,42 +45,49 @@ public void RegisterSnapshotProvider(IDependenciesSnapshotProvider snapshotProvi
...
                 SnapshotProviderUnloading?.Invoke(this, e);
 
-                lock (_snapshotProviders)
-                {
-                    _snapshotProviders.Remove(snapshotProvider.CurrentSnapshot.ProjectPath);
-                    snapshotProvider.SnapshotRenamed -= OnSnapshotRenamed;
-                    snapshotProvider.SnapshotChanged -= OnSnapshotChanged;
-                    snapshotProvider.SnapshotProviderUnloading -= OnSnapshotProviderUnloading;
-                }
+                ImmutableInterlocked.Update(
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs]: 💡 `ImmutableInterlocked.TryRemove`

File: src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs [hunk 3]
```diff
@@ -91,20 +101,17 @@ public IDependenciesSnapshot GetSnapshot(string projectFilePath)
...
-                    snapshotProvider = _projectExportProvider.GetExport<IDependenciesSnapshotProvider>(projectFilePath);
+                snapshotProvider = _projectExportProvider.GetExport<IDependenciesSnapshotProvider>(projectFilePath);
 
-                    if (snapshotProvider != null)
-                    {
-                        RegisterSnapshotProvider(snapshotProvider);
-                    }
+                if (snapshotProvider != null)
+                {
+                    RegisterSnapshotProvider(snapshotProvider);
```
Comment (sharwell) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/Snapshot/AggregateDependenciesSnapshotProvider.cs]: ❗️ If `GetSnapshot` is called in parallel, could this result in multiple event handlers getting registered?

**Comment by drewnoakes** [comment]
This hasn't been forgotten. I am looking more deeply into timing around project renames and related bugs.

**Comment by drewnoakes** [comment]
The work needed to address the larger issues discovered (see #4474) will require significantly different changes, so I'm going to close this PR for now.

This lock will likely still go for reads, and only be present to serialise infrequent updates.

**Pull request closed without merging**